snap/snapcraft.yaml

```
@@ -69,6 +76,7 @@ apps:
 
   pmqtest:
     command: usr/bin/pmqtest
+    plugs: [ posix-mq ]
```

Is this actually using the defined slot? The names don't match.

---

This creates a plug with the same name as the interface, then you have to do a self connection on the declared slot:

 ```bash
 snap connections rt-tests
Interface        Plug                                      Slot                        Notes
cpu-control      rt-tests:cpu-control                      -                           -
mount-observe    rt-tests:mount-observe                    -                           -
posix-mq         -                                         rt-tests:posix-mq-provider  -
posix-mq         rt-tests:posix-mq                         -                           -
process-control  rt-tests:process-control                  -                           -
system-files     rt-tests:sys-kernel-debug-sched-features  -                           -
system-trace     rt-tests:system-trace                     -                           -
WARNING: There are 16 new warnings. See 'snap warnings'.
 ```

Then you connect::

```bash
sudo snap connect rt-tests:posix-mq rt-tests:posix-mq-provider
```

Then here it is:

```bash
snap connections rt-tests
Interface        Plug                                      Slot                        Notes
cpu-control      rt-tests:cpu-control                      -                           -
mount-observe    rt-tests:mount-observe                    -                           -
posix-mq         rt-tests:posix-mq                         rt-tests:posix-mq-provider  manual
process-control  rt-tests:process-control                  -                           -
system-files     rt-tests:sys-kernel-debug-sched-features  -                           -
system-trace     rt-tests:system-trace                     -                           -
WARNING: There are 16 new warnings. See 'snap warnings'.
```

---

This works differently from the `custom-device` interface, if that's what you have in mind.

---

The plug definition is obsolete. The slot defined at the top level gives the access to the whole snap. That's why the application works (access is granted) even without connecting that plug. The application needs the slot to create and open a queue, and then with the same level of access can transfer messages to/from the queue to run the tests.

Example uses are available at https://github.com/snapcore/snapd/blob/20a6539e54212cc5ae92c8ada931779407156269/interfaces/builtin/posix_mq_test.go#L139-L148

The way posix_mq interface should be used is very similar to DBus, which has better documented examples: https://snapcraft.io/docs/dbus-interface

Suggestions:
1) Sot sure if this actually isolates the access to this particular app, but it is nevertheless a good way of indicating this is the app that needs the slot:
```suggestion
    slots: [ posix-mq ]
```
2) And change the top-level slot's name to `posix-mq`.

---

@farshidtz I try these, it has the same problem that still working without being connected first.

---

There is nothing to connect. The suggestion is for a standalone slot, which makes snapd apply the right apparmor/seccomp profiles. If another snap or app wants to transfer messages from/to that same queue, then it would need to define a plug and connect to that slot. I think in this case we don't need a plug because the same app creates/opens/transfers messages on the same queue.

If the slot is defined on the top level, then all apps within the snap are granted access. If the slot is set to an app, then the access will be isolated to that app alone. You can verify that by removing the slot from `pmqtest`, adding it to another app, then checking if `pmqtest` can still make the syscall.